feat: Add Longhorn V2 Data Engine

tserong opened the pull request:
This adds a new page under Advanced describing how to enable the Longhorn v2 Data Engine, along with its prerequisites and current limitations.

A couple of questions:
- Regarding the limitations, should this be worded differently?
  - Do I need to add anything about how future versions of Longhorn and Harvester will have more features added?
  - Do I need to go into more detail about the implications of some of the limitations (e.g.: as VMs with V2 volumes can't be live migrated, so will need to be shut down and restarted when upgrading Harvester)?
- Is there enough detail in "Using the Longhorn V2 Data Engine"? Is it sufficiently obvious that the V2 Data Engine is only available for new additional volumes, and existing volumes, VM images, and VM root volumes remain using the V1 engine?

(We should also update the screenshots and description on the Multi-Disk Management page, but this also probably needs to include information about @Vicente-Cheng's LVM work, so I haven't included that as part of this PR.)

Related issue: https://github.com/harvester/harvester/issues/5274

tserong commented:
The relevant pages in the preview are:

- https://6715ee2355c555ebd227b818--harvester-preview.netlify.app/v1.4/advanced/longhorn-v2/
- https://6715ee2355c555ebd227b818--harvester-preview.netlify.app/v1.4/advanced/index#longhorn-v2-data-engine-enabled-technical-preview

bk201 commented on docs/advanced/longhorn-v2.md:
```diff
@@ -0,0 +1,55 @@
+
+   If you wish to selectively disable the Longhorn V2 data engine on some nodes (for example on nodes with less CPU and/or RAM), go to the **Hosts** screen and add the label `node.longhorn.io/disable-v2-data-engine` with value `true` to those nodes.
+
+2. Go to the **Hosts** screen and add extra disks to each node as described in [Multi-disk Management](../host/host.md#multi-disk-management). Set the additional disk's `Provisioner` to `Longhorn V2 (CSI)`.
```
We can also add some screenshots in the multidisk management page for LVM/v2 volumes.

jillian-maroket commented on docs/advanced/longhorn-v2.md:
```diff
@@ -0,0 +1,55 @@
+- Volume Encryption
+- Volume Expansion
+
+Within the context of Harvester, this means that volumes backed by the Longhorn V2 data engine must be added to virutal machines as additional disks. The boot disk of each VM still needs to come from a VM image backed by the Longhorn V1 Data Engine.
+
+VMs with Longhorn V2 volumes attached cannot be live migrated, nor is snapshot functionality available, as snapshot restore within Harvester relies on volume clone functionality.
```
Alex wants me to use "virtual machine" instead of "VM" across the board.

jillian-maroket commented on docs/advanced/longhorn-v2.md:
```diff
@@ -0,0 +1,55 @@
+
+   If you wish to selectively disable the Longhorn V2 data engine on some nodes (for example on nodes with less CPU and/or RAM), go to the **Hosts** screen and add the label `node.longhorn.io/disable-v2-data-engine` with value `true` to those nodes.
+
+2. Go to the **Hosts** screen and add extra disks to each node as described in [Multi-disk Management](../host/host.md#multi-disk-management). Set the additional disk's `Provisioner` to `Longhorn V2 (CSI)`.
```
`Provisioner` is not mentioned in [Multi-disk Management](https://docs.harvesterhci.io/v1.4/host/#multi-disk-management).

tserong replied:
> `Provisioner` is not mentioned in [Multi-disk Management](https://docs.harvesterhci.io/v1.4/host/#multi-disk-management).

I wanted to do that separately because it also needs to mention LVM as a possible provisioner.

jillian-maroket commented on docs/advanced/longhorn-v2.md:
```diff
@@ -0,0 +1,55 @@
+
+2. Go to the **Hosts** screen and add extra disks to each node as described in [Multi-disk Management](../host/host.md#multi-disk-management). Set the additional disk's `Provisioner` to `Longhorn V2 (CSI)`.
+
+3. Go to the **Advanced** > **Storage Classes** screen a new storage classs as described in [Creating a Storage Class](storageclass.md#creating-a-storageclass). As with adding disks, set the `Provisioner` to `Longhorn V2 (CSI)`.
```
`Provisioner` is not mentioned in [StorageClass](https://docs.harvesterhci.io/v1.4/advanced/storageclass/).

tserong replied:
> `Provisioner` is not mentioned in [StorageClass](https://docs.harvesterhci.io/v1.4/advanced/storageclass/).

I wanted to do that separately because it also needs to mention LVM as a possible provisioner.

jillian-maroket commented on docs/advanced/settings.md:
```diff
@@ -246,6 +246,26 @@ Harvester appends necessary addresses to user-specified `noProxy` values (for ex
+
+**Versions**: v1.4.0 and later
+
+**Definition**: Enable the Longhorn V2 data engine.
+
+Setting this to `true` will automatically load the kernel modules necessary for the Longhorn V2 data engine, and will attempt to allocate 1024 x 2MiB-sized huge pages (i.e. 2GiB of RAM) on all nodes. If you see "not enough hugepages-2Mi capacity" errors when applying this setting, wait a minute for the error to clear. If the error remains, reboot the affected node(s).
+
+Changing this setting will restart RKE2 on all nodes. This will not affect running VM workloads.
+
+If you wish to selectively disable the Longhorn V2 data engine on some nodes (for example on nodes with less CPU and/or RAM), go to the Hosts screen and add the label `node.longhorn.io/disable-v2-data-engine` with value `true` to those nodes.
```
```suggestion
**Definition**: Setting that enables and disables the Longhorn V2 Data Engine.

When set to `true`, Harvester automatically loads the kernel modules required by the Longhorn V2 Data Engine, and attempts to allocate 1024 × 2 MiB-sized huge pages (for example, 2 GiB of RAM) on all nodes. 

Changing this setting automatically restarts RKE2 on all nodes but does not affect running virtual machine workloads.

:::tip

If you encounter error messages that include the phrase "not enough hugepages-2Mi capacity", allow some time for the error to be resolved. If the error persists, reboot the affected nodes.

To disable the Longhorn V2 Data Engine on specific nodes (for example, nodes with less processing and memory resources), go to the **Hosts** screen and add the following label to the target nodes:
  
- label: `node.longhorn.io/disable-v2-data-engine`
- value: `true`

:::
```

jillian-maroket commented:
Here are my answers:

> Do I need to add anything about how future versions of Longhorn and Harvester will have more features added?
No. The "Experimental" label and how the Limitations section is worded already imply that the feature set is incomplete.

> Do I need to go into more detail about the implications of some of the limitations (e.g.: as VMs with V2 volumes can't be live migrated, so will need to be shut down and restarted when upgrading Harvester)?
Such details could be useful. We can probably reuse some content from [this section](https://docs.harvesterhci.io/v1.4/advanced/singlenodeclusters/#upgrades-and-maintenance) of the doc.

> Is there enough detail in "Using the Longhorn V2 Data Engine"? 
This section leans heavily on other doc topics (Multi-disk Management, Creating a StorageClass). Linking is the correct approach. We just need to ensure that those topics actually contain the information that the user needs.

> Is it sufficiently obvious that the V2 Data Engine is only available for new additional volumes, and existing volumes, VM images, and VM root volumes remain using the V1 engine?
We can add an intro paragraph to "Using the Longhorn V2 Data Engine" that contains these details. We can also convert the related paragraph in the "Limitations" section to an "Important" note.

tserong commented on docs/advanced/longhorn-v2.md:
```diff
@@ -0,0 +1,55 @@
+
+- The Longhorn V2 Data engine is a **Technical Preview** in Harvester v1.4.0.
+
+:::
```
> Can the V2DE be used in production environments?

Good point, I'll add that (also we're changing it from "Preview" to "Experimental" for v1.4.0)

tserong commented on docs/advanced/longhorn-v2.md:
```diff
@@ -0,0 +1,55 @@
+
+- 1 CPU core for use by the Longhorn instance-manager pod
+- 2 GiB RAM (this will be allocated as 1024 × 2GiB hugepages)
+- At least one NVMe disk for data storage
```
> Item 3: Should we say "local NVMe disk" to match what's in the [Longhorn doc](https://longhorn.io/docs/1.7.0/v2-data-engine/prerequisites/#disk)?

Yeah, probably :-) will do.

tserong commented:
Thanks @jillian-maroket, I've updated this now to include all your feedback, which I've squashed into the first commit. I've also added two other commits, one to make important info boxes, and the second to add a note about an obscure issue with disk size. I'm not 100% sure if this is the best place to put it, because the issue only occurs when testing with non-NVMe disks, and we already said that local NVMe disks are a requirement (it probably really belongs in the Longhorn docs - see my comments in the commit message at https://github.com/harvester/docs/pull/653/commits/314ff949686446a707f5cc17f4a7b8c84774fc8f)

jillian-maroket commented on docs/advanced/longhorn-v2.md:
```diff
@@ -0,0 +1,85 @@
+
+  Set the `Provisioner` of each extra disk to `Longhorn V2 (CSI)`.
+
+  :::note
+
+  Harvester sets the Longhorn disk driver to "auto" as described in [the Longhorn documentation](https://longhorn.io/docs/1.7.2/v2-data-engine/features/node-disk-support/). This means that NVMe disks will use the nvme bdev driver which provides the best performance. If testing with non-NVMe disks (e.g. SSDs), note that these will fall back to using the aio bdev driver, which requires that the dis ...
+  :::
```
```suggestion
  :::note

  Harvester sets the [Longhorn disk driver](https://longhorn.io/docs/1.7.2/v2-data-engine/features/node-disk-support/) to `auto` so that NVMe disks use the SPDK NVMe bdev driver, which provides the best performance.
  
  SSDs and other non-NVMe disks are managed using the SPDK AIO bdev driver, which requires a disk size that is an *even multiple of 4096 bytes*. Non-NVMe disks that do not meet this size requirement cannot be added.

  :::
```

jillian-maroket replied:
I added "SPDK" to the driver names because those seem to be official terms in the [SPDK doc](https://spdk.io/doc/bdev.html).

jillian-maroket reviewed (commented):
Thanks, @tserong. I made a few edits to the note you added but the doc looks fab. Should be good to go soon.
